controller/src/main/java/org/jboss/as/controller/operations/global/ReadFeatureHandler.java

```diff
@@ -0,0 +1,1179 @@
+                                } else {
+                                    refs = feature.get(REFERENCES);
+                                }
+                                if (registration.getParent() != null && registration.getParent().isFeature()) {
+                                    addReference(refs, registration.getParent());
+                                }
+                                for (RegistrationPoint regPoint : (Set<RegistrationPoint>) capReg.getRegistrationPoints()) {
+                                    ModelNode ref = new ModelNode();
+                                    ref.get(FEATURE).set(root.getSubModel(regPoint.getAddress()).getFeature());
+                                    refs.add(ref);
+                                }
+                            }
+                        }
+                    }
+                }
+                if (!registration.getRequirements().isEmpty()) {
+                    PathAddress aliasAddress = createAliasPathAddress(registration, registration.getPathAddress());
+                    for (CapabilityReferenceRecorder requirement : registration.getRequirements()) {
+                        String[] segments = requirement.getRequirementPatternSegments(null, aliasAddress);
+                        String[] dynamicElements;
+                        if (segments == null) {
+                            dynamicElements = null;
+                        } else {
+                            dynamicElements = new String[segments.length];
+                            for (int i = 0; i < segments.length; i++) {
+                                dynamicElements[i] = "$" + segments[i];
+                            }
+                        }
+                        String baseRequirementName;
+                        if (isProfile) {
+                            baseRequirementName = $PROFILE_PREFIX + requirement.getBaseRequirementName();
+                        } else {
+                            baseRequirementName = requirement.getBaseRequirementName();
+                        }
+                        ModelNode capability = new ModelNode();
+                        if (dynamicElements == null) {
+                            capability.get(NAME).set(baseRequirementName);
+                        } else {
+                            capability.get(NAME).set(RuntimeCapability.buildDynamicCapabilityName(baseRequirementName, dynamicElements));
+                        }
+                        required.add(capability);
+                    }
+                }
+                if (!required.asList().isEmpty()) {
+                    feature.get(ModelDescriptionConstants.REQUIRES).set(required);
+                }
+            }
+        }
+    }
+
+    private ImmutableManagementResourceRegistration getRootRegistration(final ImmutableManagementResourceRegistration registration) {
+        while (!PathAddress.EMPTY_ADDRESS.equals(registration.getPathAddress())) {
```

Comment:
This 'while' can/should be an 'if'. Or this method can not recurse and L823 becomes registration = registration.getParent();